src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java

```diff
@@ -291,30 +289,36 @@ static ParsedSQLCacheItem parseAndCacheSQL(Sha1HashKey key,
     private int statementPoolingCacheSize = DEFAULT_STATEMENT_POOLING_CACHE_SIZE;
 
     /** Cache of prepared statement handles */
-    private ConcurrentLinkedHashMap<Sha1HashKey, PreparedStatementHandle> preparedStatementHandleCache;
+    private ConcurrentLinkedHashMap<CityHash128Key, PreparedStatementHandle> preparedStatementHandleCache;
     /** Cache of prepared statement parameter metadata */
-    private ConcurrentLinkedHashMap<Sha1HashKey, SQLServerParameterMetaData> parameterMetadataCache;
+    private ConcurrentLinkedHashMap<CityHash128Key, SQLServerParameterMetaData> parameterMetadataCache;
     /**
      * Checks whether statement pooling is enabled or disabled. The default is set to true;
      */
     private boolean disableStatementPooling = true;
 
-    /**
-     * Find statement parameters.
-     * 
-     * @param sql
-     *            SQL text to parse for number of parameters to intialize.
-     */
-    private static int countParams(String sql) {
-        int nParams = 0;
+     /**
+      * Locate statement parameters.
+      * 
+      * @param sql
+      *          SQL text to parse for positions of parameters to intialize.
+      */
+    private static int[] locateParams(String sql) {
+        List<Integer> parameterPositions = new ArrayList<Integer>(0);
 
-        // Figure out the expected number of parameters by counting the
-        // parameter placeholders in the SQL string.
+        // Locate the parameter placeholders in the SQL string.
         int offset = -1;
-        while ((offset = ParameterUtils.scanSQLForChar('?', sql, ++offset)) < sql.length())
-            ++nParams;
+        while ((offset = ParameterUtils.scanSQLForChar('?', sql, ++offset)) < sql.length()) {
+            parameterPositions.add(offset);
+        }
 
-        return nParams;
+        // Convert to int[]
+        int[] result = new int[parameterPositions.size()];
+        int i = 0;
+        for (Integer parameterPosition : parameterPositions) {
+            result[i++] = parameterPosition;
+        }
```

Comment:
Use a LinkedList to make adding/iterating faster OR use a parralell stream with ArrayList. Not necessary to keep `int i`, replace lines 317 - 320 with the following:
`parameterPositions.stream().parallel().forEach(counter -> result[counter] = parameterPositions.get(counter));`